vioscsi/helper.c

```diff
@@ -82,53 +86,55 @@ ENTER_FN_SRB();
             }
         } else {
             RhelDbgPrint(TRACE_LEVEL_ERROR, " StorPortGetStartIoPerfParams failed srb 0x%p status 0x%x MessageNumber %d.\n", Srb, status, param.MessageNumber);
+            // FIXME
+            // Should we return on this error..?
+            // Should it be TRACE_LEVEL_FATAL..?
+            //is_done = TRUE;
         }
     }
-
+    
     srbExt = SRB_EXTENSION(Srb);
-
+    
     if (!srbExt) {
         RhelDbgPrint(TRACE_LEVEL_INFORMATION, " No SRB Extenstion for SRB 0x%p \n", Srb);
+EXIT_FN_SRB();
         return;
     }
-
+    
     MessageID = QUEUE_TO_MESSAGE(QueueNumber);
     index = QueueNumber - VIRTIO_SCSI_REQUEST_QUEUE_0;
-
+    
     if (adaptExt->reset_in_progress) {
-        RhelDbgPrint(TRACE_LEVEL_FATAL, " Reset is in progress, completing SRB 0x%p with SRB_STATUS_BUS_RESET.\n", Srb);
+        RhelDbgPrint(TRACE_LEVEL_WARNING, " Reset is in progress, completing SRB 0x%p with SRB_STATUS_BUS_RESET.\n", Srb);
         SRB_SET_SRB_STATUS(Srb, SRB_STATUS_BUS_RESET);
         CompleteRequest(DeviceExtension, Srb);
+EXIT_FN_SRB();
         return;
     }
-
-    VioScsiVQLock(DeviceExtension, MessageID, &LockHandle, FALSE);
+    VioScsiLockManager(DeviceExtension, MessageID, &LockHandle, FALSE, VIOSCSI_VQLOCKOP_LOCK);
     SET_VA_PA();
     res = virtqueue_add_buf(adaptExt->vq[QueueNumber],
         srbExt->psgl,
         srbExt->out, srbExt->in,
         &srbExt->cmd, va, pa);
-
+    
     if (res >= 0) {
+        notify = virtqueue_kick_prepare(adaptExt->vq[QueueNumber]);
         element = &adaptExt->processing_srbs[index];
         InsertTailList(&element->srb_list, &srbExt->list_entry);
         element->srb_cnt++;
-    }
-    VioScsiVQUnlock(DeviceExtension, MessageID, &LockHandle, FALSE);
-    if ( res >= 0){
-        if (virtqueue_kick_prepare(adaptExt->vq[QueueNumber])) {
-            virtqueue_notify(adaptExt->vq[QueueNumber]);
-        }
     } else {
-        virtqueue_notify(adaptExt->vq[QueueNumber]);
         ScsiStatus = SCSISTAT_QUEUE_FULL;
         SRB_SET_SRB_STATUS(Srb, SRB_STATUS_BUSY);
         SRB_SET_SCSI_STATUS(Srb, ScsiStatus);
         StorPortBusy(DeviceExtension, 10);
         CompleteRequest(DeviceExtension, Srb);
-        RhelDbgPrint(TRACE_LEVEL_FATAL, " Could not put an SRB into a VQ, so complete it with SRB_STATUS_BUSY. QueueNumber = %d, SRB = 0x%p, Lun = %d, TimeOut = %d.\n", QueueNumber, srbExt->Srb, SRB_LUN(Srb), Srb->TimeOutValue);
+        RhelDbgPrint(TRACE_LEVEL_WARNING, " Could not put an SRB into a VQ, so complete it with SRB_STATUS_BUSY. QueueNumber = %d, SRB = 0x%p, Lun = %d, TimeOut = %d.\n", QueueNumber, srbExt->Srb, SRB_LUN(Srb), Srb->TimeOutValue);
+    }
+    VioScsiLockManager(DeviceExtension, MessageID, &LockHandle, FALSE, VIOSCSI_VQLOCKOP_UNLOCK);
```

Comment:
I believe so, but probably something underlying in the virtqueue implementation too.

Reply:
>Could the early unlock be the cause for the vioscsi "Reset to device \Device\RaidPortN" + IO lockups?

Yes, under high load conditions, this could certainly be a trigger.

It's probably worth pointing out that `virtqueue_kick_prepare()` and and `virtqueue_kick()` are two different things. An actual kick includes the `virtqueue_notify()`, which executes the notification callback on the VQ.

Here is the latter from `VirtIOPCICommon.c`:

```c
void virtqueue_notify(struct virtqueue *vq)
{
    vq->notification_cb(vq);
}

void virtqueue_kick(struct virtqueue *vq)
{
    if (virtqueue_kick_prepare(vq)) {
        virtqueue_notify(vq);
    }
}
```
However, `virtqueue_kick_prepare()` - which in our case calls `virtqueue_kick_prepare_split()` - prepares the buffer and returns TRUE if the device needs to be notified. Here is the return comparator for reference:

```c
if (_vq->vdev->event_suppression_enabled) {
    return wrap_around || (bool)vring_need_event(vring_avail_event(&vq->vring), new, old);
} else {
    return !(vq->vring.used->flags & VIRTQ_USED_F_NO_NOTIFY);
}
```
...and:
```c
/* The Host uses this in used->flags to advise the Guest: don't kick me when
 * you add a buffer.  It's unreliable, so it's simply an optimization.  Guest
 * will still kick if it's out of buffers.  */
#define VIRTQ_USED_F_NO_NOTIFY	1
```
It also bears mentioning, that even when a notification callback is required it is completely up to the device as to what it will do. It may well just ignore the kicks...